### PR TITLE
Fix importing of attribute ownerships during migration

### PR DESCRIPTION
**Pull request by flyingsilverfin**
## What is the goal of this PR?

The data importer (part of TypeDB's export/import functionality) correctly includes attribute ownerships on nested relations. Nested/codependent relations are imported in one batch at the end of a data import. It appears we we previously forgot to include attribute ownerships on these nested relations.

## What are the changes implemented in this PR?

* We now also import attribute ownerships on relations that were not loaded because all role players were not present
* We make the checksum mismatch error more meaningful.

### CODE REVIEW

File: migrator/data/DataImporter.java [hunk 1]
```diff
@@ -84,7 +84,7 @@
     private final int parallelisation;
 
     private final Path dataFile;
-    private final IDMapper IDMapper;
+    private final IDMapper idMapper;
```
Comment (flyingsilverfin) [inline migrator/data/DataImporter.java:87]: rename to use proper capitalisation conventions

File: migrator/data/DataImporter.java [hunk 2]
```diff
@@ -410,16 +410,26 @@ private void importSkippedRelations() {
...
+                idMapper.put(msg.getId(), importedRelation.getIID());
+
+                int ownershipCount = 0;
+                for (DataProto.Item.OwnedAttribute ownership : msg.getAttributeList()) {
+                    Thing attribute = transaction.concepts().getThing(idMapper.get(ownership.getId()));
+                    assert attribute != null;
+                    importedRelation.setHas(attribute.asAttribute());
+                    ownershipCount++;
+                }
+                status.ownershipCount.addAndGet(ownershipCount);
```
Comment (flyingsilverfin) [inline migrator/data/DataImporter.java:423]: we must include attribute ownerships that were skipped, along with the relations that were skipped, or the data import is incomplete!

File: migrator/data/DataImporter.java [hunk 3]
```diff
@@ -111,15 +111,15 @@ public void run() {
             new ParallelImport(EntitiesAndOwnerships::new).execute();
             new ParallelImport(CompleteRelations::new).execute();
             importSkippedRelations();
-            if (!checksum.verify(status)) throw TypeDBException.of(IMPORT_CHECKSUM_MISMATCH);
+            if (!checksum.verify(status)) throw TypeDBException.of(IMPORT_CHECKSUM_MISMATCH, checksum.mismatch(status));
```
Comment (flyingsilverfin) [inline migrator/data/DataImporter.java:114]: return what mismatched to the user as a more meaningful error

File: migrator/data/DataImporter.java [hunk 4]
```diff
@@ -535,5 +545,26 @@ public boolean verify(Status status) {
...
+                mismatch += "\nRelation count mismatch: expected " + relations + ", but imported " + status.relationCount.get();
+            }
+            if (roles != status.roleCount.get()) {
+                mismatch += "\nRole count mismatch: expected " + roles + ", but imported " + status.roleCount.get();
+            }
+            if (ownerships != status.ownershipCount.get()) {
+                mismatch += "\nOwnership count mismatch: expected " + ownerships + ", but imported " + status.ownershipCount.get();
+            }
+            return mismatch;
+        }
```
Comment (flyingsilverfin) [inline migrator/data/DataImporter.java:568]: clearer message when the data imported mismatches the expected data

**Review by haikalpribadi — APPROVED** [review]

**Pull request merged**